src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -648,6 +661,69 @@ public byte[] Http2AltSvcOriginUri
             return await GetHttpConnectionAsync(request, async, cancellationToken).ConfigureAwait(false);
         }
 
+        private bool IsConnectionMaximumReached(Http2Connection[]? currentHttp2Connections) => currentHttp2Connections != null && EnableMultipleHttp2Connections && currentHttp2Connections.Length == _poolManager.Settings._maxHttp2ConnectionsPerServer;
+
+        private Http2Connection PickRandomConnection(HttpRequestMessage request, Http2Connection[] currentHttp2Connections)
+        {
+            // All connections are busy so pick a random one to wait on it.
+            int randomIndex = request.GetHashCode() % _poolManager.Settings._maxHttp2ConnectionsPerServer;
+            return currentHttp2Connections[randomIndex];
+        }
+
+        private Http2Connection? GetHttp2ConnectionAcceptingNewStreams(HttpRequestMessage request)
+        {
+            Http2Connection[]? localConnections = _http2Connections;
+
+            if (localConnections == null)
+            {
+                return null;
+            }
+
+            for (int i = 0; i < localConnections.Length; i++)
+            {
+                Http2Connection http2Connection = localConnections[i];
+
+                TimeSpan pooledConnectionLifetime = _poolManager.Settings._pooledConnectionLifetime;
+                if (http2Connection.LifetimeExpired(Environment.TickCount64, pooledConnectionLifetime))
+                {
+                    // Connection expired.
+                    http2Connection.Dispose();
+                    InvalidateHttp2Connection(http2Connection);
+                }
+                else if (!EnableMultipleHttp2Connections || http2Connection.CanAddNewStream)
+                {
+                    return http2Connection;
+                }
+            }
+
+            // Re-read _http2Connections because some connections could have been invalidated above.
+            localConnections = _http2Connections;
+            if (IsConnectionMaximumReached(localConnections))
+            {
+                return PickRandomConnection(request, localConnections!);
+            }
+
+            return null;
+        }
+
+        private void AddHttp2Connection(Http2Connection newConnection)
+        {
+            lock (SyncObj)
+            {
+                Http2Connection[]? localHttp2Connections = _http2Connections;
+                int newCollectionSize = localHttp2Connections == null ? 1 : localHttp2Connections.Length + 1;
+                Http2Connection[] newHttp2Connections = new Http2Connection[newCollectionSize];
+                newHttp2Connections[0] = newConnection;
```

Comment:
Why do we add the connection at the beginning of the list? I thought we were going to add it to the end.

Reply:
To have the search from the newest to the oldest in `GetHttp2ConnectionAcceptingNewStreams`. If we decide to change the order, we will change it here.

Reply:
I guess what I'm asking is, why did we decide to search from newest to oldest instead of oldest to newest?

I don't have a strong opinion here, but I'd like to understand the rationale.

Reply:
The assumption is that older connections tend to have longer running requests than new ones because the lack of available streams on the old connections is the only driver for creating new ones. Thus, it looks more efficiently to firstly check the newest for available streams.

Reply:
Ok.